### ISSUE TITLE
Error occurred during multiple GPU training.

### ISSUE THREAD

**Issue by is** [ultralytics/ultralytics#4530]
### Search before asking

- [x] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

I run classify train task on a A10x4 host. Training works fine on a single GPU.

When I specify "device=0,1,2,3" to use multi-GPU training, after the training finishes, the following exception occurs and the training results cannot be saved properly:
```
Traceback (most recent call last):
  File "/opt/conda/envs/cs/bin/yolo", line 33, in <module>
    sys.exit(load_entry_point('ultralytics', 'console_scripts', 'yolo')())
  File "/data/jovyan/work/cs/ultralytics/ultralytics/cfg/__init__.py", line 432, in entrypoint
    getattr(model, mode)(**overrides)  # default args from model
  File "/data/jovyan/work/cs/ultralytics/ultralytics/engine/model.py", line 380, in train
    self.model, _ = attempt_load_one_weight(str(self.trainer.best))
  File "/data/jovyan/work/cs/ultralytics/ultralytics/nn/tasks.py", line 610, in attempt_load_one_weight
    ckpt, weight = torch_safe_load(weight)  # load ckpt
  File "/data/jovyan/work/cs/ultralytics/ultralytics/nn/tasks.py", line 549, in torch_safe_load
    return torch.load(file, map_location='cpu'), file  # load
  File "/opt/conda/envs/cs/lib/python3.10/site-packages/torch/serialization.py", line 791, in load
    with _open_file_like(f, 'rb') as opened_file:
  File "/opt/conda/envs/cs/lib/python3.10/site-packages/torch/serialization.py", line 271, in _open_file_like
    return _open_file(name_or_buffer, mode)
  File "/opt/conda/envs/cs/lib/python3.10/site-packages/torch/serialization.py", line 252, in __init__
    super().__init__(open(name, mode))
FileNotFoundError: [Errno 2] No such file or directory: '/data/jovyan/work/cs/ultralytics/runs/classify/train15/weights/best.pt'
```

### Minimal Reproducible Example

```
yolo classify train data=mnist160 model=yolov8x-cls.pt epochs=3 imgsz=320 device=0,1,2,3 batch=256
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@is thanks for the bug report! I am able to reproduce this issue with the command below. I'll add a TODO to prioritize a fix.

```console
yolo train model=yolov8n-cls.pt data=imagenet100 imgsz=224 epochs=3 device=4,5,6,7

Ultralytics YOLOv8.0.161 🚀 Python-3.10.11 torch-2.0.1 CUDA:4 (NVIDIA A100-SXM4-80GB, 81229MiB)
                                                       CUDA:5 (NVIDIA A100-SXM4-80GB, 81229MiB)
                                                       CUDA:6 (NVIDIA A100-SXM4-80GB, 81229MiB)
                                                       CUDA:7 (NVIDIA A100-SXM4-80GB, 81229MiB)
engine/trainer: task=classify, mode=train, model=yolov8n-cls.pt, data=imagenet100, epochs=3, patience=50, batch=16, imgsz=224, save=True, save_period=-1, cache=False, device=(4, 5, 6, 7), workers=8, project=None, name=None, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, show=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, vid_stride=1, line_width=None, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, boxes=True, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, cfg=None, tracker=botsort.yaml, save_dir=/usr/src/ultralytics/runs/classify/train2
train /usr/src/datasets/imagenet100/train... found 100 images in 100 classes ✅ 
val /usr/src/datasets/imagenet100/val... found 100 images in 100 classes ✅ 
test None...
Overriding model.yaml nc=1000 with nc=100

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]             
  3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
  4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]             
  5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]               
  6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]           
  7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]              
  8                  -1  1    460288  ultralytics.nn.modules.block.C2f             [256, 256, 1, True]           
  9                  -1  1    458340  ultralytics.nn.modules.head.Classify         [256, 100]                    
YOLOv8n-cls summary: 99 layers, 1566388 parameters, 1566388 gradients, 3.5 GFLOPs
Transferred 156/158 items from pretrained weights
DDP command: ['/opt/conda/bin/python3', '-m', 'torch.distributed.run', '--nproc_per_node', '4', '--master_port', '45207', '/root/.config/Ultralytics/DDP/_temp_vjiqm8av139935094768304.py']
train /usr/src/datasets/imagenet100/train... found 100 images in 100 classes ✅ 
val /usr/src/datasets/imagenet100/val... found 100 images in 100 classes ✅ 
test None...
DDP info: RANK 0, WORLD_SIZE 4, DEVICE cuda:0
TensorBoard: Start with 'tensorboard --logdir .', view at http://localhost:6006/
WARNING ⚠️ TensorBoard graph visualization failure 'str' object has no attribute 'parameters'
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed ✅
train: Scanning /usr/src/datasets/imagenet100/train... 100 images, 0 corrupt: 100%|██████████| 100/100 [00:00<?, ?it/s]
albumentations: RandomResizedCrop(p=1.0, height=224, width=224, scale=(0.5, 1.0), ratio=(0.75, 1.3333333333333333), interpolation=1), HorizontalFlip(p=0.5), ColorJitter(p=0.5, brightness=[0.6, 1.4], contrast=[0.6, 1.4], saturation=[0.30000000000000004, 1.7], hue=[-0.015, 0.015]), Normalize(p=1.0, mean=(0.0, 0.0, 0.0), std=(1.0, 1.0, 1.0), max_pixel_value=255.0), ToTensorV2(always_apply=True, p=1.0, transpose_mask=False)
val: Scanning /usr/src/datasets/imagenet100/val... 100 images, 0 corrupt: 100%|██████████| 100/100 [00:00<?, ?it/s]
optimizer: AdamW(lr=0.000714, momentum=0.9) with parameter groups 26 weight(decay=0.0), 27 weight(decay=0.0005), 27 bias(decay=0.0)
Closing dataloader mosaic
Image sizes 224 train, 224 val
Using 16 dataloader workers
Logging results to .
Starting training for 3 epochs...

      Epoch    GPU_mem       loss  Instances       Size
        1/3     0.388G      1.046          1        224: 100%|██████████| 7/7 [00:01<00:00,  5.31it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 13/13 [00:00<00:00, 135.38it/s]
                   all          0       0.05

      Epoch    GPU_mem       loss  Instances       Size
        2/3     0.388G      1.016          1        224: 100%|██████████| 7/7 [00:00<00:00, 22.98it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 13/13 [00:00<00:00, 193.32it/s]
                   all       0.01       0.07

      Epoch    GPU_mem       loss  Instances       Size
        3/3      0.39G      1.019          1        224: 100%|██████████| 7/7 [00:00<00:00, 25.51it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 13/13 [00:00<00:00, 118.11it/s]
                   all          0       0.07

3 epochs completed in 0.001 hours.
Optimizer stripped from weights/last.pt, 3.2MB
Optimizer stripped from weights/best.pt, 3.2MB
Results saved to .
Traceback (most recent call last):
  File "/opt/conda/bin/yolo", line 33, in <module>
    sys.exit(load_entry_point('ultralytics', 'console_scripts', 'yolo')())
  File "/usr/src/ultralytics/ultralytics/cfg/__init__.py", line 432, in entrypoint
    getattr(model, mode)(**overrides)  # default args from model
  File "/usr/src/ultralytics/ultralytics/engine/model.py", line 380, in train
    self.model, _ = attempt_load_one_weight(str(self.trainer.best))
  File "/usr/src/ultralytics/ultralytics/nn/tasks.py", line 609, in attempt_load_one_weight
    ckpt, weight = torch_safe_load(weight)  # load ckpt
  File "/usr/src/ultralytics/ultralytics/nn/tasks.py", line 548, in torch_safe_load
    return torch.load(file, map_location='cpu'), file  # load
  File "/opt/conda/lib/python3.10/site-packages/torch/serialization.py", line 791, in load
    with _open_file_like(f, 'rb') as opened_file:
  File "/opt/conda/lib/python3.10/site-packages/torch/serialization.py", line 271, in _open_file_like
    return _open_file(name_or_buffer, mode)
  File "/opt/conda/lib/python3.10/site-packages/torch/serialization.py", line 252, in __init__
    super().__init__(open(name, mode))
FileNotFoundError: [Errno 2] No such file or directory: '/usr/src/ultralytics/runs/classify/train2/weights/best.pt'

```

**Comment by glenn-jocher** [comment]
This seems to be a save_dir issue because the save directory seems to be just '.', whereas single GPU properly identifies the save_dir in runs/train...


```console
yolo train model=yolov8n-cls.pt data=imagenet100 imgsz=224 epochs=3 device=4      
Ultralytics YOLOv8.0.161 🚀 Python-3.10.11 torch-2.0.1 CUDA:4 (NVIDIA A100-SXM4-80GB, 81229MiB)
engine/trainer: task=classify, mode=train, model=yolov8n-cls.pt, data=imagenet100, epochs=3, patience=50, batch=16, imgsz=224, save=True, save_period=-1, cache=False, device=4, workers=8, project=None, name=None, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, show=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, vid_stride=1, line_width=None, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, boxes=True, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, cfg=None, tracker=botsort.yaml, save_dir=/usr/src/ultralytics/runs/classify/train2
train /usr/src/datasets/imagenet100/train... found 100 images in 100 classes ✅ 
val /usr/src/datasets/imagenet100/val... found 100 images in 100 classes ✅ 
test None...
Overriding model.yaml nc=1000 with nc=100

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]             
  3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
  4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]             
  5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]               
  6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]           
  7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]              
  8                  -1  1    460288  ultralytics.nn.modules.block.C2f             [256, 256, 1, True]           
  9                  -1  1    458340  ultralytics.nn.modules.head.Classify         [256, 100]                    
YOLOv8n-cls summary: 99 layers, 1566388 parameters, 1566388 gradients, 3.5 GFLOPs
Transferred 156/158 items from pretrained weights
TensorBoard: Start with 'tensorboard --logdir /usr/src/ultralytics/runs/classify/train2', view at http://localhost:6006/
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed ✅
train: Scanning /usr/src/datasets/imagenet100/train... 100 images, 0 corrupt: 100%|██████████| 100/100 [00:00<?, ?it/s]
albumentations: RandomResizedCrop(p=1.0, height=224, width=224, scale=(0.5, 1.0), ratio=(0.75, 1.3333333333333333), interpolation=1), HorizontalFlip(p=0.5), ColorJitter(p=0.5, brightness=[0.6, 1.4], contrast=[0.6, 1.4], saturation=[0.30000000000000004, 1.7], hue=[-0.015, 0.015]), Normalize(p=1.0, mean=(0.0, 0.0, 0.0), std=(1.0, 1.0, 1.0), max_pixel_value=255.0), ToTensorV2(always_apply=True, p=1.0, transpose_mask=False)
val: Scanning /usr/src/datasets/imagenet100/val... 100 images, 0 corrupt: 100%|██████████| 100/100 [00:00<?, ?it/s]
optimizer: AdamW(lr=0.000714, momentum=0.9) with parameter groups 26 weight(decay=0.0), 27 weight(decay=0.0005), 27 bias(decay=0.0)
Image sizes 224 train, 224 val
Using 8 dataloader workers
Logging results to /usr/src/ultralytics/runs/classify/train2
Starting training for 3 epochs...

      Epoch    GPU_mem       loss  Instances       Size
        1/3     0.401G      1.043          4        224: 100%|██████████| 7/7 [00:00<00:00,  7.72it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 4/4 [00:00<00:00, 71.47it/s]
                   all          0       0.03

      Epoch    GPU_mem       loss  Instances       Size
        2/3      0.38G      1.038          4        224: 100%|██████████| 7/7 [00:00<00:00, 35.03it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 4/4 [00:00<00:00, 193.70it/s]
                   all       0.02       0.06

      Epoch    GPU_mem       loss  Instances       Size
        3/3     0.384G      1.013          4        224: 100%|██████████| 7/7 [00:00<00:00, 36.47it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 4/4 [00:00<00:00, 28.81it/s]
                   all       0.03       0.09

3 epochs completed in 0.001 hours.
Optimizer stripped from /usr/src/ultralytics/runs/classify/train2/weights/last.pt, 3.2MB
Optimizer stripped from /usr/src/ultralytics/runs/classify/train2/weights/best.pt, 3.2MB
Results saved to /usr/src/ultralytics/runs/classify/train2
```

**Comment by is** [comment]
It seems directly setting save_dir to an absolute path does not solve the problem. Are there any other quick fixes? I can try locally.

**Comment by glenn-jocher** [comment]
@is great news 😃! Your original issue may now be resolved ✅ in Multi-GPU DDP training fix PR #4544. 

To get this update:
- **[Git](https://github.com/ultralytics/ultralytics)** – Run `git pull` from within your `ultralytics/` directory or run `git clone https://github.com/ultralytics/ultralytics` again
- **[Pip](https://pypi.org/project/ultralytics/)** – Update with `pip install -U ultralytics`
- **[Notebooks](https://github.com/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb)** – Check out the updated notebooks  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/ultralytics)** – Run `sudo docker pull ultralytics/ultralytics:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and letting us know. Please confirm if this update fixes the issue for you, and don't hesitate to report any other **issues** you find or **feature requests** you may have. Happy training with YOLOv8 🚀!
